### ISSUE TITLE
Imagej Macro Runner on a specific channel

### ISSUE THREAD

**Issue by bpavie** [qupath/qupath#68]
Dear,

I tried to perform a simple ImageJ Macro on a specific channel, but even if the channel is selected, it looks like the **ImageJ Macro Runner** is sending the whole color image, so my script that was tailored for DAB only failed.
Would it be possible to select the channel as it is when we send the region using **Extensions>ImageJ>Send Region to ImageJ** ?

Also, it looks like that if I can send a rescaled region by a certain factor using **Extensions>ImageJ>Send Region to ImageJ** (e.g. by 4) , I cannot use the same downscale factor using the **ImageJ Macro Runner**, which popup a message telling me the image is too big, requiring a smaller downscale factor (looks like a bug)

Otherwise, thanks again for this great piece of software!
Best
Benjamin

**Comment by petebankhead** [comment]
Thanks Benjamin, and I'm glad the macro functionality is useful!

I thought I knew the reason for the issues you are seeing, but upon quickly looking at the code I realise that I do not.  There is an enigmatic statement [here](https://github.com/qupath/qupath/blob/master/qupath-extension-ij/src/main/java/qupath/imagej/plugins/ImageJMacroRunner.java#L368) that color transforms are ```// Not supported in batch mode, so disable option to avoid confusion``` although I don't recall why...

There's also a [hard-coded limit](https://github.com/qupath/qupath/blob/master/qupath-extension-ij/src/main/java/qupath/imagej/plugins/ImageJMacroRunner.java#L229) whereby the macro runner rejects an image with more than 5000x5000 pixels... although the wisdom of this specific limit may also be questionable.

I'll look into it, but at the minute it looks like both issues represent 'intended behavior, albeit not desired behavior (by the developer or anyone else)'.  Not sure what the right word for that is.

In the meantime, depending upon how happy you are with Groovy/Java/the ImageJ API, it would be possible to create a Groovy script to run in QuPath that grabs regions from the image, converts them into ```ImagePlus``` objects for ImageJ, performs whatever processing is required using ImageJ (or even OpenCV or other dependencies if you prefer), and optionally sends back results as the ```PathObjects``` that QuPath requires.

There is considerably more effort involved in setting this up for the first time and learning the main methods required (IntelliJ is more or less essential to get auto-complete and link up to the source code), but has the reward of giving you far more ability to customize the analysis and how the results are returned.  I have used this approach a lot.

If you would like to try this out, the code within [```ImageJMacroRunner.java```](https://github.com/qupath/qupath/blob/master/qupath-extension-ij/src/main/java/qupath/imagej/plugins/ImageJMacroRunner.java) may help to get started, or see the ```Estimate_background_values.groovy``` script included as supplementary material with the [bioRxiv preprint](http://biorxiv.org/content/early/2017/03/06/099796).

**Comment by bpavie** [comment]
Thanks Pete for your answer and pointing to the scripts, I will have a look at them.
Actually, the limit size shouldn't be more set to the Java array limit size which is **Integer.MAX_VALUE** no (and match also the maximum memory) ?

**Comment by petebankhead** [comment]
Possibly... Java should take care of any requests exceeding the ```Integer.MAX_VALUE```, but the maximum memory should probably be involved in the calculation somewhere.

I faintly recall adding a limit because I kept accidentally crashing/hanging by requesting regions that were much too large, and the macro runner doesn't do any kind of smart breaking-a-large-region-into-tiles.  So I wanted to protect me from myself.

It's complicated by the fact that large continuous regions of memory need to be found for the arrays and QuPath has no idea how much memory the macro is going to require (i.e. how many image duplicates it will make).  And maybe sometimes the macro will be run in parallel.

I still tend towards having a fixed limit, just a much bigger one.  And preferably one implemented as a static field somewhere.  Then it could be modified by a script in an emergency - or possibly even within the macro runner dialog itself.  I definitely would agree that hard-coding a relatively small limit in the middle of a method wasn't the optimal way...

**Comment by petebankhead** [comment]
Closing this due to lack of activity, and because much has changed on the path to v0.2.0. Hopefully the size issue is mostly addressed.

There is definitely still room for improvement with the macro runner, but the current behavior is limited by the current design... alas, I think QuPath may need a new macro runner in a future version. We may create a new issue if we have a solid proposal for what this should look like - in the meantime, we can use the forum if needed to discuss requests.

**Comment by mertesdorfj** [comment]
Hello!

I wanted to ask if there are any updates on this potential ImageJ macro runner extension. I have a similar case as @bpavie and would be interested in finding a way to send a single channel (DAB) from QuPath to ImageJ. Also, @bpavie if you found a way to get this to work, I´d be very happy about any input. Thank you!

Best, 
Julia

**Comment by petebankhead** [comment]
Hi Julia, I think this is also your question?
https://forum.image.sc/t/qupath-analysis-of-nerve-cells-sending-a-single-channel-dab-to-imagej/93593
The forum is really a better place for this discussion, and more people who can help are likely to see it there.

**Comment by mertesdorfj** [comment]
Dear Pete,

yes that was also my question. I wasn´t sure if it´s better to post the question here or on Image.sc, thus I tried both. Thanks for your suggestions on the forum thread!

Best,
Julia
